### DISCUSSION TITLE
Consider restructuring CDM files

### DISCUSSION THREAD

**Discussion by korikuzma** [cancervariants/metakb#514] [Ideas]
Right now, we have one giant json dump that contains the following: 

```python
class TransformedData(BaseModel):
    """Define model for transformed data"""

    statements_evidence: list[Statement] = Field(
        [], description="Statement objects for evidence records"
    )
    statements_assertions: list[
        VariantTherapeuticResponseStudyStatement
        | VariantPrognosticStudyStatement
        | VariantDiagnosticStudyStatement
    ] = Field([], description="Statement objects for assertion records")
    categorical_variants: list[CategoricalVariant] = []
    variations: list[Allele] = []
    genes: list[MappableConcept] = []
    therapies: list[MappableConcept | TherapyGroup] = []
    conditions: list[MappableConcept] = []
    methods: list[Method] = []
    documents: list[Document] = []
```

We should consider a better structure for our CDM files. For example, should we use referenced or dereferenced? What makes the most sense for loading concepts into the Graph DB?

**Comment by korikuzma** [comment]
@cancervariants/metakb-maintainers would love your thoughts on this. It will be blocking me soon.

**Comment by jsstevenson** [comment]
@korikuzma what are some deficiencies w/r/t the current structure?I think it'd be fine if you wanted to enref stuff for compactness but I don't think it would significantly alter how we ingest the data?

**Reply by korikuzma** [reply]
The current structure works fine, but I think we could organize things better. 

Two possibilities I was thinking of based off of MOA:
* Dereferenced: Single file with all the statement types (`Statements`, `VariantTherapeuticResponseStudyStatements`, `VariantPrognosticStudyStatements`, and `VariantDiagnosticStudyStatement`)
  * Keys for each type or list containing all. I like the former better 
* Referenced: Single file for each GKS concept type

**Reply by korikuzma** [reply]
Maybe it would be good to give both referenced and dereferenced options?

**Reply by jsstevenson** [reply]
Yeah, I guess it depends on where we envision this output being used. For MetaKB I think we'd like to assume a specific form of input but it probably doesn't matter how it's organized (one file is always nicer but not mandatory or anything).

**Reply by katie-perry** [reply]
I think if it's not too difficult, both options would be amazing. It would fit a variety of use-cases. I do also think that the current CDM structure is suitable for our needs. The organization change would probably make ingesting some things easier, but I can't think of anything that I've been unable to find or had to write a bunch of extra code to find in the current structure. I'm going to take a look at the code I wrote for VarCat and double-check, though.

**Reply by korikuzma** [reply]
> Dereferenced: Single file with all the statement types (Statements, VariantTherapeuticResponseStudyStatements, VariantPrognosticStudyStatements, and VariantDiagnosticStudyStatement) 

So the Study Statements will have evidence lines that contain references to `Statements`. For example CIViC AID 6 is a `VariantTherapeuticResponseStudyStatement` with 6 supporting `Statement`s in its `hasEvidenceLines` field. Did we want these evidence items to be referenced (iris) or in this case still dereference and be okay with duplicate data?

**Reply by jsstevenson** [reply]
> So the Study Statements will have evidence lines that contain references to Statements. For example CIViC AID 6 is a VariantTherapeuticResponseStudyStatement with 6 supporting Statements in its hasEvidenceLines field. Did we want these evidence items to be referenced (iris) or in this case still dereference and be okay with duplicate data?

8/29 planning call: we'll try to quantify how often this happens (assertions that are ingestible but have evidence lines that aren't ingestible) and move forward from there

**Comment by jsstevenson** [comment]
One point I will add. Right now, we try to do upload all data in a single transaction, ie if anything fails then everything gets rolled back. If we had more data to add, we might not want to do that -- instead, we might want to be transacting based around associated entities. IE, instead of

```
<! start transaction>
upload gene 1...
upload gene 2...
upload gene 3...
upload therapy 1...
upload therapy 2...
<! failure, rollback>
```

We could instead do
```
<! start transaction>
upload therapy 1...
upload gene 1...
upload variation 1...
upload disease 1...
upload statement 1...
<! complete transaction>
<! start transaction>
upload therapy 2...
upload gene 2...
upload variation 2...
upload disease 2...
<! failure, rollback just this transaction>
<! start transaction>
upload therapy 3...
```

and so on. To do this, I think it'd be easier if the CDM supported some way of iterating by statement. Either embed full versions of all entity objects under the statement (redundant, but simple), or just include IRIs and then make the other entity subgroups keyed by ID rather than simply be arrays

so 
```json
{
  "statements": [
    {
      "id": "civic.aid:12345",
      "type": "TherapeuticResponseProposition",
      "subjectVariation": {
        "id": "civic.mpid:33",
        "type": "CategoricalVariant"
      },
      "condition": {
        "id": "civic.did:11",
        "type": "Condition",
        "name": "colon cancer"
      }
    }
  ]
}
```

or

```json
{
  "statements": [
    {
      "id": "civic.aid:12345",
      "type": "TherapeuticResponseProposition",
      "subjectVariation": "civic.mpid:33",
      "condition": "civic.did:11"
    }
  ],
  "subjectVariations": {
    "civic.mpid:33": {
      "id": "civic.mpid:33",
      "type": "CategoricalVariant"
    }
  },
  "conditions": {
    "civic.did:11": {
      "id": "civic.did:11",
      "type": "Condition",
      "name": "colon cancer"
    }
  }
}
```

caveat: I haven't thought about how evidence lines work here, maybe that gets tricky

**Reply by jsstevenson** [reply]
Also, we obviously could figure out a way to do this without changing the CDM at all.... like, when we load the CDM JSON, we could just restructure the entity arrays as dicts etc. So this doesn't necessarily require any changes to the CDM. Just something I was thinking about.
